Frontend routes

MMojtaba opened the pull request:
Added functions:
-Create mock schedule
-Retrieve schedule of all teachers or this teacher's schedule
-Set and get teacher pereferences.

MMojtaba committed:
Added create mock schedule row and retrieving all schedule

MMojtaba committed:
Added documentation to get schedule and made mock schedule to create multiple rows of a schedule

MMojtaba committed:
Added my schedule endpoint. Added check for authorization to schedule/list.

MMojtaba committed:
Added email to schedule table since it's used as the unique identifer and changed schedule code accordingely.

MMojtaba committed:
Added teacher preference: my, list, and update functions.

MMojtaba committed:
Added teacher preference: my, list, and update functions.

MMojtaba committed:
Made it so it auto creates preferences when accessing my when none exists. Added documentation for teacher pref.

MMojtaba committed:
Removed some comments

KjartanE commented on api/src/routes/schedule.routes.ts:
```diff
@@ -1,4 +1,100 @@
+
+    try
+    {
+        const response = await scheduleController.my(token);
```
Instead of creating a new function called `my' which collects user data on its own. Could it init a user.controller to get the user data and then separately have a function in schedule that gets all schedules based from a user email or id. That way we maintain any data invloved in parsing JWT is maintained inside the auth or user controller

MMojtaba replied:
Not sure if I understood it corrctly; I'm using user.controller to get user data now, and I don't pass authtoken to my (or other functions in the schedule controller). Let me know if I should change something.

KjartanE commented on api/src/controllers/schedule.controller.ts:
```diff
@@ -0,0 +1,89 @@
+     */
+    async list(authToken: string): Promise<ISchedule[]>
+    {
+        let decoded_email = '';
```
Or if you are only using the JWT to validate the USERS role is ADMIN. THen create a new function in Auth controller, that takes in the JWT token and returns if user is admin or not

MMojtaba replied:
lol good point, it's way cleaner now. I'm doing authentication in routes now. 
Should be done.

KjartanE commented on api/src/controllers/teacherpref.controller.ts:
```diff
@@ -0,0 +1,111 @@
+     */
+    async my(authToken: string): Promise<ITeacherPref>
+    {
+        let user = await getUser(authToken);
```
Right idea here, but lets utilize this more in each endpoint. In addition, makes more sense to me if this user GET is done in the endpoint so that this function does not need to know anything about JWT

KjartanE replied:
ie: funtion only accepts user object

MMojtaba replied:
Should be done.

KjartanE commented on api/src/models/schedule.model.ts:
```diff
@@ -0,0 +1,46 @@
+const Schema = mongoose.Schema;
+
+//term, course, section, instructor, capacity, location, days, start, end
+
```
Remove old code, no need to keep around

KjartanE commented on api/src/controllers/schedule.controller.ts:
```diff
@@ -0,0 +1,89 @@
+ * @export
+ * @class ScheduleController
+ */
+export class ScheduleController 
```
Also moving forward, lets write tests for the files we are creating. I understand these are subject to change. But if we have tests written and are confirming object types and data. THen updating the response or input will break tests but is easy to update.

MMojtaba replied:
Added tests for schedule controller and teacherpref controller.

MMojtaba committed:
For schedule and teacherpref: using a function from usercontroller to retrieve user. Using a function from auth
to check if user is authorized. Authentication (checking for admin and verifying authToken) is done in routes now, and authtoken is not passed to controllers.

MMojtaba committed:
Added tests for schedule controller

MMojtaba committed:
Fixed failing test

MMojtaba committed:
Seperated schedule controller tests into different describe sections

MMojtaba committed:
Added testse for teacher pref controller

KjartanE reviewed (approved):
🐈  Great work thanks for doing all those refactors
